src/router.zig

```diff
@@ -108,98 +66,176 @@ pub fn Router(comptime Context: type, comptime routes: []const Route(Context)) R
+                            return route.handler(context, response, request, object.params[0..object.param_count]);
                     }
                 },
             }
         }
     }.serve;
 }
 
+/// Will automatically convert router captures to function parameter types
+pub fn wrap(comptime Context: type, comptime handler: anytype) Route(Context).Handler {
+    const info = @typeInfo(@TypeOf(handler));
+    if (info != .Fn)
+        @compileError("router.wrap expects a function type");
+
+    const function_info = info.Fn;
+    if (function_info.is_generic)
+        @compileError("Cannot create handler wrapper for generic function");
+    if (function_info.is_var_args)
+        @compileError("Cannot create handler wrapper for variadic function");
+
+    if (function_info.args[0].arg_type.? != Context)
+        @compileError("Argument 0 of a HandlerFn must be Context");
+    if (function_info.args[1].arg_type.? != *Response)
+        @compileError("Argument 1 of a HandlerFn must be *Response");
+    if (function_info.args[2].arg_type.? != Request)
+        @compileError("Argument 2 of a HandlerFn must be Request");
+
+    const capture_args_info = function_info.args[3..];
+    var capture_arg_types: [capture_args_info.len]type = undefined;
+    for (capture_args_info) |arg, i| {
+        capture_arg_types[i] = arg.arg_type.?;
+    }
+    const CaptureArgs = std.meta.Tuple(&capture_arg_types);
+
+    const X = struct {
+        fn wrapper(ctx: Context, res: *Response, req: Request, params: []const trie.Entry) anyerror!void {
+            var capture_args: CaptureArgs = undefined;
+
+            std.debug.assert(params.len == capture_args.len); // Number of captures must equal the number of extra parameters in the Handler function
+
+            if (capture_args.len == 0) return handler(ctx, res, req);
+
+            comptime var arg_index = 0;
+            inline while (arg_index < capture_args.len) : (arg_index += 1) {
+                capture_args[arg_index] = switch (@TypeOf(capture_args[arg_index])) {
+                    []const u8 => params[arg_index].value,
+                    //?[]const u8 => if (params.len > 0) params[0].value else null,
+                    else => |ArgType| switch (@typeInfo(ArgType)) {
+                        .Int => std.fmt.parseInt(ArgType, params[arg_index].value, 10) catch {
+                            return res.notFound();
+                        },
+                        .Optional => |child| if (@typeInfo(child) == .Int)
+                            std.fmt.parseInt(ArgType, params[arg_index].value, 10) catch null
+                        else
+                            @compileError("Unsupported optional type " ++ @typeName(child)),
+                        else => @compileError("Unsupported type " ++ @typeName(ArgType)),
+                    },
+                };
+            }
```

Comment:
Perhaps we shouldn't do any type conversion at all, and always provide the capture as a slice of bytes to the user, and have them parse it as an integer or whatever instead. Let me know if you think otherwise.

Reply:
I think that's sensible, as there are some many different ways to parse a parameter. Knowing that the parameter was passed in still removes a lot of boilerplate.

What do you think about matching the captures? Right now I have it so that each additional argument is a capture, but we could make additional parameters require a struct, e.g.:

```zig
// What I have currenctly

fn post(ctx: Ctx, res: Response, req: Request, post: usize, message: []const u8) !void {
    // do response...
}

// With no type conversions and single parameter for captures

fn post(ctx: Ctx, res: Response, req: Request, captures: struct {
    post: []const u8,
    message: []const u8,
}) !void {
    // do response...
}
```

This would also allow matching the parameter names to the struct's fields.

Reply:
I like the idea of requiring a struct for multiple captures similar to what we had before https://github.com/Luukdegram/apple_pie/commit/5399084b774a609b6058b700ee5777f5d4376733. It makes the usage less error-prone for the end user who may mess up the capture order and we gain the ability to format the function signature in multiple ways for free.